File: aihound/core/redactor.py
"""Secret value masking and redaction."""

from __future__ import annotations

import re

# Known credential prefixes and their display names
KNOWN_PREFIXES = {
    "sk-ant-": "Anthropic",
    "sk-ant-ort": "Anthropic Refresh",
    "sk-ant-oat": "Anthropic Access",
    "sk-": "OpenAI/Generic",
    "ghp_": "GitHub PAT (classic)",
    "gho_": "GitHub OAuth",
    "ghu_": "GitHub User-to-Server",
    "ghs_": "GitHub Server-to-Server",
    "github_pat_": "GitHub PAT (fine-grained)",
    "xoxb-": "Slack Bot",
    "xoxp-": "Slack User",
    "xoxa-": "Slack App",
    "AKIA": "AWS Access Key",
    "AIza": "Google API Key",
    "ya29.": "Google OAuth Access",
}


def mask_value(value: str, show_full: bool = False) -> str:
    """Mask a credential value, preserving known prefixes and last 4 chars.

    Examples:
        sk-ant-oat01-abc...xF2q
        ghp_abc1...9xYz
        ***REDACTED***  (for short values)
    """
    if show_full:
        return value

    if not value or len(value) <= 8:
        return "***REDACTED***"

    # Find the longest matching known prefix
    matched_prefix = ""
    for prefix in sorted(KNOWN_PREFIXES.keys(), key=len, reverse=True):
        if value.startswith(prefix):
            matched_prefix = prefix
            break

    if matched_prefix:
        # Show prefix + a few more chars + ... + last 4
        extra = min(4, len(value) - len(matched_prefix) - 4)
        if extra > 0:
            preview_start = value[: len(matched_prefix) + extra]
        else:
            preview_start = value[: len(matched_prefix)]
        return f"{preview_start}...{value[-4:]}"

    # No known prefix: show first 6 + ... + last 4
    return f"{value[:6]}...{value[-4:]}"


# Pre-compiled patterns for redact_line
_PREFIX_PATTERN = "|".join(re.escape(p) for p in sorted(KNOWN_PREFIXES.keys(), key=len, reverse=True))
_REDACT_TOKEN_RE = re.compile(
    r"(?<![A-Za-z0-9_\-])((?:" + _PREFIX_PATTERN + r")[A-Za-z0-9_\-./+=]{16,})"
)
_REDACT_CONTEXT_RE = re.compile(
    r"""(?ix)
    (
      (?:api[_-]?key|token|secret|password|passwd|auth[a-z_-]*|bearer)\s*[=:]\s*
    | export\s+[A-Z_][A-Z0-9_]*\s*=\s*
    | -H\s+["']?Authorization:\s*Bearer\s+
    | -H\s+["']?x-api-key:\s*
    | --api-key\s+
    | --token\s+
    )
    ["']?([A-Za-z0-9_\-./+=]{20,})["']?
""",
)


def redact_line(line: str) -> str:
    """Redact any credential values found in a line of text.

    Uses KNOWN_PREFIXES to find and mask known token prefixes, and also
    redacts values after common assignment/header patterns like ``=``,
    ``export VAR=value``, ``-H Authorization: Bearer``, etc.

    Returns the line with credential values replaced by masked forms.
    """
    # Pass 1: Replace known-prefix tokens
    def _mask_prefix_match(m: re.Match) -> str:
        return mask_value(m.group(1))

    result = _REDACT_TOKEN_RE.sub(_mask_prefix_match, line)

    # Pass 2: Replace context-based credential values
    def _mask_context_match(m: re.Match) -> str:
        prefix_part = m.group(1)
        value_part = m.group(2)
        return prefix_part + mask_value(value_part)

    result = _REDACT_CONTEXT_RE.sub(_mask_context_match, result)

    return result


def identify_credential_type(value: str) -> str | None:
    """Try to identify what kind of credential a value is based on its prefix."""
    for prefix, name in sorted(KNOWN_PREFIXES.items(), key=lambda x: len(x[0]), reverse=True):
        if value.startswith(prefix):
            return name
    return None

File: Other Versions/Go/core/redactor.go
package core

import (
	"regexp"
	"sort"
	"strings"
	"sync"
)

// knownPrefix maps a credential prefix to its human-readable type name.
type knownPrefix struct {
	Prefix string
	Name   string
}

// knownPrefixes is sorted longest-first for correct matching.
var knownPrefixes []knownPrefix

func init() {
	raw := []knownPrefix{
		{"sk-ant-ort", "Anthropic Refresh Token"},
		{"sk-ant-oat", "Anthropic Access Token"},
		{"sk-ant-", "Anthropic API Key"},
		{"sk-", "OpenAI/Generic API Key"},
		{"github_pat_", "GitHub PAT (fine-grained)"},
		{"ghp_", "GitHub PAT (classic)"},
		{"gho_", "GitHub OAuth Token"},
		{"ghu_", "GitHub User-to-Server Token"},
		{"ghs_", "GitHub Server-to-Server Token"},
		{"xoxb-", "Slack Bot Token"},
		{"xoxp-", "Slack User Token"},
		{"xoxa-", "Slack App Token"},
		{"AKIA", "AWS Access Key"},
		{"AIza", "Google API Key"},
		{"ya29.", "Google OAuth Access Token"},
	}
	// Sort by prefix length descending for longest-prefix-first matching
	sort.Slice(raw, func(i, j int) bool {
		return len(raw[i].Prefix) > len(raw[j].Prefix)
	})
	knownPrefixes = raw
}

// findMatchingPrefix returns the prefix string and type name for a value,
// or empty strings if no known prefix matches.
func findMatchingPrefix(value string) (string, string) {
	for _, kp := range knownPrefixes {
		if len(value) >= len(kp.Prefix) && value[:len(kp.Prefix)] == kp.Prefix {
			return kp.Prefix, kp.Name
		}
	}
	return "", ""
}

// MaskValue masks a credential value, preserving known prefixes and last 4 chars.
// If showFull is true, the value is returned unchanged.
func MaskValue(value string, showFull bool) string {
	if showFull {
		return value
	}

	if value == "" || len(value) <= 8 {
		return "***REDACTED***"
	}

	prefix, _ := findMatchingPrefix(value)
	if prefix != "" {
		// Show prefix + up to 4 extra chars + ... + last 4
		extra := len(value) - len(prefix) - 4
		if extra > 4 {
			extra = 4
		}
		var previewStart string
		if extra > 0 {
			previewStart = value[:len(prefix)+extra]
		} else {
			previewStart = value[:len(prefix)]
		}
		return previewStart + "..." + value[len(value)-4:]
	}

	// No known prefix: show first 6 + ... + last 4
	return value[:6] + "..." + value[len(value)-4:]
}

// IdentifyCredentialType tries to identify what kind of credential a value is
// based on its prefix. Returns empty string if unrecognized.
func IdentifyCredentialType(value string) string {
	_, name := findMatchingPrefix(value)
	return name
}

// GetKnownPrefixes returns the list of known credential prefixes,
// sorted longest-first (for regex alternation construction).
func GetKnownPrefixes() []string {
	out := make([]string, 0, len(knownPrefixes))
	for _, kp := range knownPrefixes {
		out = append(out, kp.Prefix)
	}
	return out
}

// redactLineRe matches known-prefix tokens in a line of text.
// Built once on first call to RedactLine.
var (
	redactLineRe   *regexp.Regexp
	redactValueRe  *regexp.Regexp
	redactLineOnce sync.Once
)

func initRedactRegex() {
	redactLineOnce.Do(func() {
		prefixes := GetKnownPrefixes()
		escaped := make([]string, len(prefixes))
		for i, p := range prefixes {
			escaped[i] = regexp.QuoteMeta(p)
		}
		prefixAlt := strings.Join(escaped, "|")
		// Match known-prefix tokens (prefix + 16+ alphanum chars)
		redactLineRe = regexp.MustCompile(`((?:` + prefixAlt + `)[A-Za-z0-9_\-./+=]{16,})`)
		// Match values after = signs and similar assignment patterns
		redactValueRe = regexp.MustCompile(`(?i)((?:api[_-]?key|token|secret|password|passwd|auth[a-z_-]*|bearer)\s*[=:]\s*["']?)([A-Za-z0-9_\-./+=]{20,})`)
	})
}

// RedactLine masks known credential patterns in a line of text intended for
// notes or log output. Known-prefix tokens are replaced with prefix + "***".
// Values after common assignment patterns (e.g. "token=...") are replaced
// with "***REDACTED***".
func RedactLine(line string) string {
	initRedactRegex()

	// Pass 1: Redact known-prefix tokens
	line = redactLineRe.ReplaceAllStringFunc(line, func(match string) string {
		prefix, _ := findMatchingPrefix(match)
		if prefix != "" {
			return prefix + "***"
		}
		if len(match) > 6 {
			return match[:6] + "***"
		}
		return "***REDACTED***"
	})

	// Pass 2: Redact values after assignment patterns
	line = redactValueRe.ReplaceAllString(line, "${1}***REDACTED***")

	return line
}
